src/compiler/performanceCore.ts

```diff
@@ -84,7 +84,7 @@ namespace ts {
                                         performance.mark(end);
                                     }
                                     performance.measure(name, start, end);
-                                    if (end = "__performance.measure-fix__") {

```

Comment:
Hm... Should we have a lint for this?

Reply:
Or a `--strictBoolean` (or `--pedanticBoolean`?) flag that requires all conditions be `boolean` (including `!`). It wouldn't catch `if (f = true)`, but it also would mean you'd have to write a lot of `if (obj !== null && obj !== undefined)` instead of `if (obj)` because `!!obj` would also be a violation...

Reply:
`while (next = read())` seems like a sane pattern, so this feels more like a heuristic than a compiler error.  Having said that, I would probably personally use such a mode, assuming it had a (totally reasonable) exception for `!!expr`.

Reply:
I mean, mostly just using `=` in an `if` without a reference to the assigned variable within the `if`. (Since I know we do this intentionally in relationship checking, but the difference there is we actually reference the checked variable)